core/frontend/src/main/scala/io/udash/routing/UrlChangeProvider.scala

```diff
@@ -2,41 +2,100 @@ package io.udash.routing
 
 import io.udash.core.{Url, Window}
 import org.scalajs.dom
+import org.scalajs.dom.Element
 
-/**
-  * Provides information about current URL.
-  */
+import scala.scalajs.js
+
+/** Provides information about current URL. */
 trait UrlChangeProvider {
-  /** Changes URL fragment following # to passed [[io.udash.core.Url]]. */
+  /** Changes the URL part representing the frontend routing state. */
   def changeFragment(url: Url): Unit
 
-  /** Returns URL fragment following #. */
+  /** Returns the URL part representing the frontend routing state. */
   def currentFragment: Url
 
-  /** Registers listener for URL fragment change. */
-  def onFragmentChange(callback: (Url) => Unit): Unit
+  /** Registers listener for URL changes. */
+  def onFragmentChange(callback: Url => Unit): Unit
 
-  /** Changes whole URL. */
+  /** Changes the whole URL. */
   def changeUrl(url: String): Unit
 }
 
-object WindowUrlChangeProvider extends UrlChangeProvider {
-
+/** Used for routing based on the URL part following # sign. */
+object WindowUrlFragmentChangeProvider extends UrlChangeProvider {
   import dom.document
 
-  private var callbacks: List[Url => Unit] = List()
-
-  override def onFragmentChange(callback: Url => Unit): Unit = {
-    callbacks = callback :: callbacks
-  }
-
-  def currentFragment: Url = Url(document.location.hash.stripPrefix("#"))
+  private val callbacks: js.Array[Url => Unit] = js.Array()
 
   Window.onFragmentChange(() => {
```

Comment:
We only ever use `Window` to store a single callback - maybe we could just add an event listener to `DomWindow` here and encapsulate it better (or use `window.document` and get rid of it)?